### PR TITLE
validate ego vehicle role name

### PR DESCRIPTION
**Pull request by makaveli10**
- validate ego vehicle role name with CARLA server and available ros bridge role names

### CODE REVIEW

File: notebooks/CARLA 💜 Foxglove demo.ipynb [hunk 1]
```diff
@@ -216,6 +216,52 @@
...
+   "source": [
+    "ego_vehicle_role_name = [\"ego\", \"hero\", \"ego_vehicle\", \"hero0\", \"hero1\", \"hero2\", \"hero3\", \"hero4\", \"hero5\", \"hero6\", \"hero7\", \"hero8\", \"hero9\"]\n",
+    "used_role_name = []\n",
+    "\n",
+    "def get_new_role_name(role_name):\n",
+    "    new_rolename = random.choice(ego_vehicle_role_name)\n",
+    "    while new_rolename == role_name or new_rolename in used_role_name:\n",
+    "        if len(used_role_name) == len(ego_vehicle_role_name):\n",
+    "            raise Exception(f\"All available role_names for ego vehicle in ROS bridge: {ego_vehicle_role_name}\\\n",
+    "                            already exist in the simulation.\")\n",
```
Comment (jpc) [inline notebooks/CARLA 💜 Foxglove demo.ipynb:233]: The `\` in Python strings does not strip the spaces at the beginning of the next line so you get strange long gaps in the output:
```
Given role_name: ego_vehicle already exists in simulation...                         selecting a different one!
```

I would stick to a single line here even if it is a bit long. (or `+` two strings in two lines)

File: notebooks/CARLA 💜 Foxglove demo.ipynb [hunk 2]
```diff
@@ -216,6 +216,52 @@
...
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ego_vehicle_role_name = [\"ego\", \"hero\", \"ego_vehicle\", \"hero0\", \"hero1\", \"hero2\", \"hero3\", \"hero4\", \"hero5\", \"hero6\", \"hero7\", \"hero8\", \"hero9\"]\n",
+    "used_role_name = []\n",
```
Comment (jpc) [inline notebooks/CARLA 💜 Foxglove demo.ipynb:226]: This is some global state but inside the function you actually check it every time against the CARLA server. It looks like you could simplify the function quite a bit if you removed this.

It looks to me you could change `get_new_role_name` to create a copy of the possible names list, iterate over CARLA actors to remove the names that are already used, and then use the first one that's left (or choose one at random). The `validate_rolename` would not be required anymore.

**Review by jpc — CHANGES_REQUESTED** [review]

File: notebooks/CARLA 💜 Foxglove demo.ipynb (hunk 2 above)
Comment (makaveli10) [reply notebooks/CARLA 💜 Foxglove demo.ipynb:226]: makes sense. thanks

File: notebooks/CARLA 💜 Foxglove demo.ipynb [hunk 3]
```diff
@@ -237,7 +258,13 @@
     "# Let CarlaViz know this vehicle is the ego vehicle.\n",
     "# Vehicles controlled by the user are commonly differenciated\n",
     "# in CARLA by setting the attribute role_name to ego.\n",
-    "blueprints_vehicle.set_attribute('role_name', 'ego_vehicle')\n",
+    "role_name = 'ego_vehicle'\n",
+    "role_name = validate_rolename(role_name)\n",
+    "if role_name==None:\n",
+    "    raise Exception(\"All available role names already exist in the simulation.\")\n",
+    "print(f\"Validated role_name: {role_name}\")\n",
```
[the remark this answers is not in the document]
Comment (makaveli10) [reply notebooks/CARLA 💜 Foxglove demo.ipynb:265]: sure, not breaking anything.

File: notebooks/CARLA 💜 Foxglove demo.ipynb [hunk 4]
```diff
@@ -216,6 +216,27 @@
     "world.apply_settings(settings);"
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "ego_vehicle_role_name = [\"ego\", \"hero\", \"ego_vehicle\", \"hero0\", \"hero1\", \"hero2\", \"hero3\", \"hero4\", \"hero5\", \"hero6\", \"hero7\", \"hero8\", \"hero9\"]\n",
```
[the remark this answers is not in the document]
Comment (makaveli10) [reply notebooks/CARLA 💜 Foxglove demo.ipynb:225]: yes, sure can do that. Also, we need to extend the role_names with the bridge as well.

[the remark this answers is not in the document]
Comment (zoq) [reply notebooks/CARLA 💜 Foxglove demo.ipynb:225]: True, the ros-bridge needs to be updates as well.

[the remark this answers is not in the document]
Comment (makaveli10) [reply notebooks/CARLA 💜 Foxglove demo.ipynb:225]: so, the bridge creates an instance of ```EgoVehicle``` only if the given role_name exist within the list the bridge has.

[the remark this answers is not in the document]
Comment (zoq) [reply notebooks/CARLA 💜 Foxglove demo.ipynb:225]: Yeah, so I guess just bump the number of role names someone can use, and keep it simple e.g. `hero0, hero1, ... , heron`.

[the remark this answers is not in the document]
Comment (zoq) [reply notebooks/CARLA 💜 Foxglove demo.ipynb:225]: I guess we could also go for `col0, col1, coln` to brand it a little bit?

[the remark this answers is not in the document]
Comment (makaveli10) [reply notebooks/CARLA 💜 Foxglove demo.ipynb:225]: yeah sounds good!

File: notebooks/CARLA 💜 Foxglove demo.ipynb [hunk 5]
```diff
@@ -13,7 +13,11 @@
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
+   "metadata": {
+    "vscode": {
+     "languageId": "python"
+    }
+   },
```
Comment (jpc) [inline notebooks/CARLA 💜 Foxglove demo.ipynb:20]: This metadata is not really useful but makes diffs harder to read. It should be removed automatically if you run `python cleanup-notebook.py notebooks/CARLA\ 💜\ Foxglove\ demo.ipynb` before committing.

File: notebooks/CARLA 💜 Foxglove demo.ipynb [hunk 6]
```diff
@@ -219,7 +243,41 @@
...
+    "    for actor in actors:\n",
+    "        if actor.attributes['role_name'] in available_rolenames:\n",
+    "            available_rolenames.remove(actor.attributes['role_name'])\n",
+    "    \n",
+    "    # all role names are taken\n",
+    "    if not len(available_rolenames):\n",
+    "        return None\n",
+    "    if role_name not in available_rolenames:\n",
+    "        role_name = random.choice(available_rolenames)\n",
+    "    return role_name"
```
Comment (jpc) [inline notebooks/CARLA 💜 Foxglove demo.ipynb:270]: 💜👍

**Review by jpc — CHANGES_REQUESTED** [review]

**Pull request merged**